src/NUnitFramework/framework/Internal/Execution/WorkItemQueue.cs

```diff
@@ -244,8 +271,11 @@ public WorkItem Dequeue()
 
 
                 // Dequeue our work item
-                WorkItem work;
-                while (!_innerQueue.TryDequeue(out work)) { };
+                WorkItem work = null;
+                while (work == null)
+                    foreach (var q in _innerQueues)
+                        if (q.TryDequeue(out work))
+                            break;
```

Comment:
@CharliePoole I am fine with merging if @ChrisMaddock or @jnm2 don't have time for a review. All of the code looks good to me, but something about this is bothering me. It is a gut reaction and I can't put my finger on it.

My thinking is that if there is nothing in any of the queues, work will always be null and this go into a hard spin. The break will just break out of the foreach. Am I incorrect? Shouldn't this just be something like,

```C#
foreach (var q in _innerQueues)
    if (q.TryDequeue(out work))
        break;

if(work == null)
    continue;
```

Which will allow another try at the Spin/Wait code above?

Or am I misunderstanding the Spin/Wait code above and it can never get here if there is nothing in any of the queues? If so, why do we need the while?

This is what concerns me about the lockless code that was added, it is very hard to understand all of the moving parts when you haven't looked at it in awhile.